File: backend/app/routers/category.py
from flask import Blueprint, request, jsonify
from app.services.category_service import CategoryService
from app.utils.jwt import jwt_required
from app.exceptions.customer_exceptions import NotFoundException, BadRequestException

category_bp = Blueprint('category', __name__)

@category_bp.route('/categories', methods=['GET'])
@jwt_required()
def get_categories():
    """Get all categories"""
    try:
        categories = CategoryService.get_all_categories()
        return jsonify({
            "data": categories,
            "message": "Categories retrieved successfully"
        }), 200
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories/tree', methods=['GET'])
@jwt_required()
def get_category_tree():
    """Get categories in tree structure"""
    try:
        tree = CategoryService.get_category_tree()
        return jsonify({
            "data": tree,
            "message": "Category tree retrieved successfully"
        }), 200
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories/<int:category_id>', methods=['GET'])
@jwt_required()
def get_category_by_id(category_id):
    """Get a specific category by ID"""
    try:
        category = CategoryService.get_category_by_id(category_id)
        return jsonify({
            "data": category,
            "message": "Category retrieved successfully"
        }), 200
    except NotFoundException as e:
        return jsonify({"message": str(e)}), 404
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories/<int:category_id>/children', methods=['GET'])
@jwt_required()
def get_category_children(category_id):
    """Get direct children of a category"""
    try:
        children = CategoryService.get_children(category_id)
        return jsonify({
            "data": children,
            "message": "Category children retrieved successfully"
        }), 200
    except NotFoundException as e:
        return jsonify({"message": str(e)}), 404
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories/<int:category_id>/descendants', methods=['GET'])
@jwt_required()
def get_category_descendants(category_id):
    """Get all descendants of a category"""
    try:
        descendants = CategoryService.get_descendants(category_id)
        return jsonify({
            "data": descendants,
            "message": "Category descendants retrieved successfully"
        }), 200
    except NotFoundException as e:
        return jsonify({"message": str(e)}), 404
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories', methods=['POST'])
@jwt_required(allowed_roles=[1, 2])  # Only admins and teachers can create categories
def create_category():
    """Create a new category"""
    try:
        data = request.get_json()
        
        if 'display_name' not in data:
            return jsonify({"message": "Display name is required"}), 400
            
        category = CategoryService.create_category(
            display_name=data['display_name'],
            parent_id=data.get('parent_id')
        )
        
        return jsonify({
            "data": category,
            "message": "Category created successfully"
        }), 201
    except NotFoundException as e:
        return jsonify({"message": str(e)}), 404
    except BadRequestException as e:
        return jsonify({"message": str(e)}), 400
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories/<int:category_id>', methods=['PUT'])
@jwt_required(allowed_roles=[1, 2])  # Only admins and teachers can update categories
def update_category(category_id):
    """Update a category"""
    try:
        data = request.get_json()
        update_data = {}
        
        if 'display_name' in data:
            update_data['display_name'] = data['display_name']
            
        if 'parent_id' in data:
            update_data['parent_id'] = data['parent_id']
            
        category = CategoryService.update_category(
            category_id=category_id,
            data=update_data
        )
        
        return jsonify({
            "data": category,
            "message": "Category updated successfully"
        }), 200
    except NotFoundException as e:
        return jsonify({"message": str(e)}), 404
    except BadRequestException as e:
        return jsonify({"message": str(e)}), 400
    except Exception as e:
        return jsonify({"message": str(e)}), 500

@category_bp.route('/categories/<int:category_id>', methods=['DELETE'])
@jwt_required(allowed_roles=[1, 2])  # Only admins and teachers can delete categories
def delete_category(category_id):
    """Delete a category"""
    try:
        recursive = request.args.get('recursive', 'false').lower() == 'true'
        
        result = CategoryService.delete_category(category_id, recursive)
        return jsonify(result), 200
    except NotFoundException as e:
        return jsonify({"message": str(e)}), 404
    except BadRequestException as e:
        return jsonify({"message": str(e)}), 400
    except Exception as e:
        return jsonify({"message": str(e)}), 500

File: frontend/src/api/category.js
import request from '@/utils/http/axios'

/**
 * Category API - 分类相关接口
 * 对应 backend/app/routes/category.py
 */
export const categoryApi = {
    /**
     * 获取所有分类
     * GET /api/category/categories
     * @returns {Promise<{data: Array, message: string}>}
     */
    getAllCategories() {
        return request.get('/api/category/categories');
    },

    /**
     * 获取分类树结构
     * GET /api/category/categories/tree
     * @returns {Promise<{data: Array, message: string}>}
     */
    getCategoryTree() {
        return request.get('/api/category/categories/tree');
    },

    /**
     * 获取指定分类信息
     * GET /api/category/categories/:id
     * @param {number} categoryId - 分类ID
     * @returns {Promise<{data: Object, message: string}>}
     */
    getCategoryById(categoryId) {
        return request.get(`/api/category/categories/${categoryId}`);
    },

    /**
     * 获取指定分类的直接子分类
     * GET /api/category/categories/:id/children
     * @param {number} categoryId - 分类ID
     * @returns {Promise<{data: Array, message: string}>}
     */
    getCategoryChildren(categoryId) {
        return request.get(`/api/category/categories/${categoryId}/children`);
    },

    /**
     * 获取指定分类的所有后代分类
     * GET /api/category/categories/:id/descendants
     * @param {number} categoryId - 分类ID
     * @returns {Promise<{data: Array, message: string}>}
     */
    getCategoryDescendants(categoryId) {
        return request.get(`/api/category/categories/${categoryId}/descendants`);
    },

    /**
     * 创建分类
     * POST /api/category/categories
     * @param {Object} data - 分类数据
     * @param {string} data.display_name - 分类名称（必填）
     * @param {number} [data.parent_id] - 父分类ID（可选）
     * @returns {Promise<{data: Object, message: string}>}
     */
    createCategory(data) {
        return request.post('/api/category/categories', data);
    },

    /**
     * 更新分类
     * PUT /api/category/categories/:id
     * @param {number} categoryId - 分类ID
     * @param {Object} data - 更新数据
     * @param {string} [data.display_name] - 分类名称
     * @param {number} [data.parent_id] - 父分类ID
     * @returns {Promise<{data: Object, message: string}>}
     */
    updateCategory(categoryId, data) {
        return request.put(`/api/category/categories/${categoryId}`, data);
    },

    /**
     * 删除分类
     * DELETE /api/category/categories/:id
     * @param {number} categoryId - 分类ID
     * @param {boolean} [recursive=false] - 是否递归删除子分类
     * @returns {Promise<{message: string}>}
     */
    deleteCategory(categoryId, recursive = false) {
        return request.delete(`/api/category/categories/${categoryId}?recursive=${recursive}`);
    }
};

export default categoryApi; 